### ISSUE TITLE
Deprecate `EnableDualStack`

### ISSUE THREAD

**Issue by bleggett** [istio/istio#52186]
**Describe the feature request**

In `istiod` and other spots + tests we have the flag `EnableDualStack`.

This flag should be entirely deprecated.

- K8S decides whether we are dual stack or not, not Istio - there's little benefit in us having our own flag for it.
- If `Istio` needs a flag, it should be `EnableV6=yes/no`, and nothing else - `Istio` should simply support V6 (or not).
- Istio should always support more-than-one-IP for all K8S resources that support multiple IPs - this has nothing to do with "dualstack", but in too many places we have used "dualstack" as a global proxy flag for "might have more than one IP" which is wrong.
 
Any code that doesn't fit into the above assertions is very likely simply wrong/a bug.

So desired endstate is

- `EnableV6=yes/no` as the only global flag.
- Implicit support for N+1 IPs in K8S resources (where we do not -> bug).
- Respecting K8S `ipFamily/ipFamilyPolicy` on resources that that have it (where we do not -> bug).
- Mimicking K8S `ipFamily/ipFamilyPolicy` on Istio-specific resources that that require it.
 
 We are not that far from doing this I think, but opening this to track removal of the flag.
 
**Describe alternatives you've considered**

**Affected product area (please put an X in all that apply)**

[ ] Ambient
[ ] Docs
[x] Dual Stack
[ ] Installation
[ ] Networking
[ ] Performance and Scalability
[ ] Extensions and Telemetry
[ ] Security
[ ] Test and Release
[ ] User Experience
[ ] Developer Infrastructure

**Affected features (please put an X in all that apply)**

[ ] Multi Cluster
[ ] Virtual Machine
[ ] Multi Control Plane

**Additional context**

**Comment by keithmattix** [comment]
+1 to this direction; when dualstack was more experimental in k8s (and istio) I think this layout made more sense, but I think it's time.

**Comment by bleggett** [comment]
Yep - the main thing is to get our integ tests running with dualstack (e.g. https://github.com/istio/istio/pull/51872 for ambient).

Once they are we will be as confident as we can reasonably be, and can drop the flag.

**Comment by hzxuzhonghu** [comment]
I donot think we should deprecate a experimental featire flag but make it default. AFAIK most people in China just deplyed a ipv4 cluster

**Comment by kfaseela** [comment]
cc @leosarra

**Comment by bleggett** [comment]
> I donot think we should deprecate a experimental featire flag but make it default. AFAIK most people in China just deplyed a ipv4 cluster

Correct. If someone wants an IPv4-only cluster they should set `ipv6=false` like we do in ambient:

https://github.com/istio/istio/blob/master/manifests/charts/istio-cni/values.yaml#L56

We don't need `DualStack=true/false` and everywhere in the code we reference it we arguably have a bug, so it shouldn't be/can't be the actual end-user UX for enabling/disabling IPv6 support.

There are two end-state scenarios here:

1. Want IPv4 only? 
 -> `ipv6=false`

2. Anything else?
 -> Set nothing, it's a kubernetes config, not an Istio config, and we mostly just confuse ourselves by saying "dualstack support" when we mean "bugs around handling N+1 IPs on resources" or "bugs around respecting K8S dualstack config".

**Comment by bleggett** [comment]
> I think it makes sense to remove the dualstack flag because a lot of the code that was originally gated behind it is now always being executed. However, adding an IPv6 flag for sidecar as a replacement would require some effort, as the current logic simply defaults to whatever address K8s provides first (default family of cluster).


It might be smarter to decouple sidecar v6 and ambient v6 support and stability levels - there is less baggage to undo, and a large chunk of the concern for breakage is with sidecars specifically.


Controlplane -> should not have global "dualstack" toggles or flags - just `v6=true/false
Dataplane > sidecar may have a "dualstack" toggle, eventually to go away in favor of the CP setting.
Ambient -> Should respect the CP setting, no "dualstack" flag.

**Comment by hzxuzhonghu** [comment]
Remove `EnableDualStack ` doenot mean eliminating the bugs around it.  The bugs are around multi addresses changes. 

>Want IPv4 only?
> -> ipv6=false

This is just a replacement of EnableDualStack?

**Comment by bleggett** [comment]
> Remove `EnableDualStack ` doenot mean eliminating the bugs around it. The bugs are around multi addresses changes.
> 
> > Want IPv4 only?
> > -> ipv6=false
> 
> This is just a replacement of EnableDualStack?

To quote @leosarra 

> I think it makes sense to remove the dualstack flag because a lot of the code that was originally gated behind it is now always being executed.

There is also the problem of `single-stack` V6 - do you set `EnableDualStack=true` to get single-stack v6?

K8S determines whether DualStack is enabled, not Istio. Whether a K8S cluster supports dualstack or not has nothing to do with whether Istio provisions resources with V6 addresses, and that's the endstate (and the UX) I want to get to.

**Comment by hzxuzhonghu** [comment]
>I think it makes sense to remove the dualstack flag because a lot of the code that was originally gated behind it is now always being executed.

It is not the reason of the env, it is people coding wrongly.  Why we need to protect with the env, actually it is because these code are unsafe and experimental, maybe rarely used in production unlike the more common features like routing, lb.
To remove a flag used to protect such unsafe code, what would people think? We have strict standard about how to elvolve the features.

**Comment by leosarra** [comment]
> It is not the reason of the env, it is people coding wrongly. Why we need to protect with the env, actually it is because these code are unsafe and experimental, maybe rarely used in production unlike the more common features like routing, lb. To remove a flag used to protect such unsafe code, what would people think? We have strict standard about how to elvolve the features.

I feel that the code now executed outside the flag was likely deemed safe by its author otherwise it would be a mistake not to put it under an experimental flag.

My proposal isn't to always run the dualstack codepaths. Rather, I'm more inclined towards removing the user-exposed flag and replacing it with internal logic. This logic would allow the DS codepaths to run if Istio is deployed on a dualstack cluster, and prevent them otherwise.
Currently, it seems the flag is used forusers to tell us if they're using a dualstack cluster, which seems unnecessary. If they set the flag wrong then Istio will missbehave as seen in DS (IPv6-default) clusters where the liveness probes will fail.
I would prefer for Istio to determine this by its own, simplifying the user experience.
